Change S(0) to S.Zero etc

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Replaced S(0) with S.Zero and so on.

Motivation:

```
In [3]: %timeit s = S(0)
693 ns ± 1.27 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit s = S.Zero
37.8 ns ± 0.759 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

I have (more or less) blindly changed instances of S(0), S(1), S(-1) and S(1)/2. There are multiple instances of - S.One but apart from a few, typically, when it was `(-S(1))**` I've left them.

The change I've made is replacing S(-1)/2 with -S.Half.

```
%timeit s = S.NegativeOne/2
2.1 µs ± 53.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit s = -S.Half
427 ns ± 7.15 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

I've tried to avoid docstrings, but it is not impossible that some automatic replacement has happened.

#### Other comments
Other values one can consider to include and their approximate occurrence in the code are:
-1/2 : S.NegativeHalf ~1700
sqrt(2): S.SqrtTwo ~2000
sqrt(3): S.SqrtThree ~1300
1/3: S.Third ~500
1/4: S.Quarter ~250
2: S.Two ~500
3: S.Three ~1900
4: S.Four ~150
3/2: S.ThreeHalves ~1600 (This is where most of the S(3) comes from)
3/4: S.ThreeQuarters ~100 (This is where most of the S(3) comes from)
1/sqrt(2), sqrt(2)/2: S.ReciprocalSqrtTwo ~200

But I guess that adding more predefined constants also slows down the general access a bit?

For sqrt:
```
In [8]: %timeit s = sqrt(2)
1.15 µs ± 1.17 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
compared to ~40 ns to looking it up.

For 3/2:
```
In [9]: %timeit s = S(3)/2
2.58 µs ± 60.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [10]: %timeit s = S.Half*3
1.64 µs ± 35.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
so here is a potential speed-up just by rewriting.

It can be noted that many instances are in the tests, but still beneficial to clean up every now and then.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscargus committed:
Change S(0) to S.Zero etc

oscargus committed:
Removed S-expressions from relationals

oscargus committed:
Replaced -S.Infinity with S.NegativeInfinity

smichr commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -538,7 +538,7 @@ def test_exgaussian():
     # Note: actual_output simplifies to expected_output.
     # Ideally cdf(X)(z) would return expected_output
-    # expected_output = (erf(sqrt(2)*(l*s**2 + m - z)/(2*s)) - 1)*exp(l*(l*s**2 + 2*m - 2*z)/2)/2 - erf(sqrt(2)*(m - z)/(2*s))/2 + S(1)/2
+    # expected_output = (erf(sqrt(2)*(l*s**2 + m - z)/(2*s)) - 1)*exp(l*(l*s**2 + 2*m - 2*z)/2)/2 - erf(sqrt(2)*(m - z)/(2*s))/2 + S.Half
```
ok in comment

smichr commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1117,9 +1117,9 @@ def test_unrad_slow():
 @XFAIL
 def test_unrad_fail():
-    # this only works if we check real_root(eq.subs(x, S(1)/3))
+    # this only works if we check real_root(eq.subs(x, S.One/3))
```
ok

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -198,7 +198,7 @@ def test_reduce_abs_inequalities():
 def test_reduce_inequalities_general():
     assert reduce_inequalities(Ge(sqrt(2)*x, 1)) == And(sqrt(2)/2 <= x, x < oo)
-    assert reduce_inequalities(PurePoly(x + 1, x) > 0) == And(S(-1) < x, x < oo)
+    assert reduce_inequalities(PurePoly(x + 1, x) > 0) == And(-1 < x, x < oo)
```
FYI: `-1 < x` will become `x > -1` whereas `S.NegativeOne < x` will remain unchanged. But apparently (since tests pass) `And` makes the args canonical.

smichr replied:
Oh, I thought tests were passing but they aren't. So I imagine that this is what caused the failure.

oscargus replied:
Yes, this was causing a bit of the failures.

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1045,7 +1045,7 @@ def test_is_convergent():
     assert Sum(eq, (x, 1, 2)).is_convergent() is S.true
     assert Sum(1/(x**3), (x, 1, oo)).is_convergent() is S.true
-    assert Sum(1/(x**(S(1)/2)), (x, 1, oo)).is_convergent() is S.false
+    assert Sum(1/(x**(S.Half)), (x, 1, oo)).is_convergent() is S.false
```
`(S.Half)`->`S.Half` globally

oscargus replied:
Except that it may be an argument to a function. I think I have fixed most though.

smichr commented on sympy/core/tests/test_match.py:
```diff
@@ -472,10 +472,10 @@ def test__combine_inverse():
-    assert Add._combine_inverse(oo*I, oo*I) == S(0)
-    assert Add._combine_inverse(x*oo, x*oo) == S(0)
-    assert Add._combine_inverse(-x*oo, -x*oo) == S(0)
+    assert Add._combine_inverse(oo, oo) == S.Zero
```
` == S.Zero\n` -> ` is S.Zero\n` globally

oscargus replied:
Good way to find bugs. Turned out the the stats module used `sum` which returns an `int` if everything summed up is 0.

smichr commented on sympy/integrals/intpoly.py:
```diff
@@ -1025,20 +1025,20 @@ def point_sort(poly, normal=None, clockwise=True):
 
     def compare(a, b):
-        if a.x - center.x >= S.Zero and b.x - center.x < S.Zero:
+        if a.x - center.x >= 0 and b.x - center.x < 0:
```
If you make this change you force sympy to sympify the value. Why did you make these changes?

oscargus replied:
There was a period of time where I was under the impression that using numbers was actually faster in relations. Let me double-check if it is true here.

oscargus replied:
Not true here. Will revert.

oscargus replied:
This was the observation I did:

```
%timeit 0 < y
10.8 µs ± 970 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit S.Zero < y
29.7 µs ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

but it cannot be applied here.

oscarbenjamin replied:
I would usually prefer to use `a < b` then `a - b < 0` (for clarity rather than performance, or possibly to handle infinities)

oscargus replied:
Yeah, but now the purpose was performance. :-) (Without reducing readability.) Not sure that the total result is very striking in terms of performance though... But at least a few things should be better.

smichr commented on sympy/integrals/meijerint.py:
```diff
@@ -1247,7 +1247,7 @@ def _int0oo(g1, g2, x):
     >>> from sympy.abc import s, t, m
     >>> from sympy import meijerg, S
-    >>> g1 = meijerg([], [], [-S(1)/2, 0], [], s**2*t/4)
+    >>> g1 = meijerg([], [], [-S.Half, 0], [], s**2*t/4)
```
Perhaps it is better to leave docstring examples unchanged.

oscargus replied:
Yes, that is the idea. But I might have missed or pressed "Replace all"... Thanks for finding it!

oscarbenjamin commented:
> But I guess that adding more predefined constants also slows down the general access a bit?

I don't see why it would. These are just attribute/dict lookups which should have `O(1)` access regardless of dict size.

I assume that you are finding these examples by grepping for `S(...)` which won't pick up on something like `3*x` even though the 3 here still needs to be sympified. Testing a bit I find that sometimes it is faster to use a regular int though:
```julia
In [1]: S.Four = S(4)                                                                                                             

In [2]: %timeit exp(S.Four)                                                                                                       
1.09 µs ± 6.43 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %timeit exp(S(4))                                                                                                         
3.24 µs ± 29.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit exp(4)                                                                                                            
642 ns ± 7.95 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

oscargus committed:
Partial fixes

oscargus commented:
> 
> 
> > But I guess that adding more predefined constants also slows down the general access a bit?
> 
> I don't see why it would. These are just attribute/dict lookups which should have `O(1)` access regardless of dict size.

OK! I thought that maybe the access time increased with dict size.

> I assume that you are finding these examples by grepping for `S(...)` which won't pick up on something like `3*x` even though the 3 here still needs to be sympified. Testing a bit I find that sometimes it is faster to use a regular int though:

Correct, plain grep.

Yes, I've found the same, so it is quite hard to know when to do what. Probably for rational numbers it will help, but then, one can also go for Rational directly:

```
%timeit s =S.One/3
4.62 µs ± 140 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit s = Rational(1, 3)
579 ns ± 12.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

(Although clearly, accessing the singleton dict would be like 40 ns.)

Also

```
%timeit s = S.Infinity
87.6 ns ± 4.16 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

%timeit s = oo
41.9 ns ± 2.4 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

So all the constants should really be better off accessed directly (although it is convenient to only have to import `S`).

I've noted that using plain numbers is faster in matrices and relationals, but then, in some tests, sometimes the relationals must be formed using `S`, at least the first number it seems.

It seems like `is_zero` is clearly faster than `== S.Zero`, `== 0`, `is S.Zero`, etc. At least if it is a symbol or sympified constant. For expressions it varies:

```
%timeit exp(x) == 0
2.91 µs ± 150 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit exp(x).is_zero
871 ns ± 17.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit exp(x) == S.Zero
2.24 µs ± 141 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit exp(x) is S.Zero
702 ns ± 25.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit S.Zero is S.Zero
138 ns ± 1.87 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

%timeit S.Zero.is_zero
90.4 ns ± 8.48 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Still, I wouldn't expect `exp(x) is S.Zero` to actually try to evaluate before deciding. So I've changed a bit of == 0 to `is_zero` in the code. Thinking about it in the tests as well since if it still evaluates to S.Zero, the actual test seems to be faster.

But there is a quite large number of things that could be changed so (I think) once I get the tests to pass again I may be happy for now.

oscarbenjamin commented:
> So I've changed a bit of == 0 to `is_zero` in the code.

These are not equivalent e.g.:
```julia
In [1]: e = cos(x)**2+sin(x)**2-1                                                                                                 

In [2]: e.is_zero                                                                                                                 

In [3]: e.is_zero == None                                                                                                         
Out[3]: True

In [4]: e == 0                                                                                                                    
Out[4]: False
```
In SymPy `==` is for structural equality. For the most part (see #17498 for an exception) Basic instances compare equal with `==` if they are of the same type and have the same args (args are also compared with `==`) i.e.:
```python
    def __eq__(self, other):
        return type(self) == type(other) and self.args == other.args
```
On the other hand `is_zero` is part of the assumptions system and uses ternary logic. Exactly when `is_zero` will be True/False rather than None is not well defined since it just depends how clever the assumptions checking code is.

oscargus commented:
Yeah, I know. Hence, it is probably quite often better to check with `is_zero` in the code:
 
```
sin(0, evaluate=False).is_zero
Out[12]: True

a = Add(1, -1, evaluate=False)

a.is_zero
Out[14]: True

sin(a).is_zero
Out[15]: False

sin(a) == 0
Out[16]: False

sin(0, evaluate=False) == 0
Out[17]: False

sin(0, evaluate=False) is S.Zero
Out[18]: False
```

(Now, there is clearly an issue with `sin.is_zero`, but if it checked with `arg.is_zero` it would be OK. Hence, it makes sense to check with that rather than == 0 as is often done now. That `is_zero` is not good enough in many situations is another thing... This is on the current state of this PR, so not sure if it is an issue here or not. The main problem I've encountered so far is that I've found some functions that doesn't sympify its arguments... But we'll see what comes out in the end.)

Other interesting observations include:

```
In [19]: x = Symbol('x')

In [20]: %timeit s = x/7
4.86 µs ± 138 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [21]: %timeit s = x*Rational(1, 7)
2.82 µs ± 100 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [22]: x/7 == x*Rational(1, 7)
Out[22]: True
```

I've not gone that far, but expressions like `3*I/2` are rewritten as `I*Rational(3, 2)`.

oscargus committed:
More rewrites

oscargus commented:
Here is maybe a better example. Consider the line:

https://github.com/sympy/sympy/blob/9c900ebf41bc3f11fb66a66ed14fbdb4acaab299/sympy/calculus/util.py#L1316

Here, it seems to me that `== 0` is the worst option (since `other` is sympified).

```
In [66]: four = S(4)

In [67]: %timeit four % 2 is S.Zero
1.38 µs ± 67.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [68]: %timeit four % 2 == 0
1.83 µs ± 282 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [69]: %timeit (four % 2).is_even
1.36 µs ± 35.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [70]: %timeit four.is_even
433 ns ± 25 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [71]: M = Symbol('M', integer=True, even=True)

In [72]: %timeit M % 2 is S.Zero
4.21 µs ± 217 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [73]: %timeit M % 2 == 0
4.65 µs ± 201 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [74]: %timeit M.is_even
428 ns ± 23.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [75]: %timeit (M % 2).is_zero
4.36 µs ± 329 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Still, there may be expressions where the eval methods are not good enough (as in the `sin` case above). However, will `== 0` actually solve this? It seems like the only time one should use `== 0` is if the argument to check is not sympified (e.g. if it is the result of e.g. `len`).

oscargus committed:
Everything pass locally

oscargus commented:
Since it very much seems like `Rational(5, 4)` is more efficient compared to `S(5)/4`, I'd suggest that the docstrings are updated where relevant to reflect that. Although not critical, it is better to show "good practice".

oscargus committed:
Fixed some issues

oscargus committed:
Fixed test failures

oscargus commented:
There are quite a lot of places where one can/should replace `== int` with `is S(int)` to make sure that the result is a SymPy `Integer`. I've left that our here though...

oscargus commented:
The tests pass! I'd be happy for quick reviews (merges...) considering the potential for merge conflicts here.

I'm sure that looking at the code shows examples of code which can be improved (like how to write comparisons, `a > b` or `a - b > 0`), but please focus on the changes in this PR. :-) (I've found lots of more or less weird code to possibly be improved, but if I did that I would never finish...)

oscarbenjamin commented:
Does this make anything noticeably faster?

smichr commented:
See also [XKCD](https://xkcd.com/1205/).

smichr commented:
I have already reviewed all the non-test files. I have mixed feelings about this, however, since except at the very lowest level the relative time savings seem very small. And we don't want to spend time working on things that don't make a difference. Decisions like these should be made at the automated checking level if they are important. And if they're not...then we just don't worry about them.

oscargus commented:
> Does this make anything noticeably faster?

Hard to say. Not on a general level, but there are clearly operations that are faster. (But then, there are other things that require a lot of time so it get drowned in the noise.) I _think_ that the `spin`(?)-tests in `quantum` that has several thousands of instantiations using `S` is one or a few seconds faster (of about 30 in total). 

I think the main benefit is really that it shows better practice how to write efficient code (using Rational instead of S(X)/Y), proper checking using `is`, the bug fix for `probability`, and some other inconsistencies in formatting. So no harm really but some benefits.

In the longer run, I plan to add at least `S.NegativeHalf` and maybe `S.SqrtTwo`. And slowly covert tests and code checks to use `is S.Zero` or ìs_zero` etc instead of `== 0`, both for efficiency and correctness (right now there are methods that may sometimes return SymPy numbers, sometimes Python numbers, and I guess that at least all SymPy code that could return symbolic expressions shouldn't return Python numbers).

I'd be tempted to change the `sstr`-printer to not print `S(2)/3`, but `Rational(2, 3)`. Not sure why this was done originally (`S` shorter than `Rational`?).

sylee957 commented:
There are some other concerns of exceeding 78-line limits easily if every library code are rewritten, and most of the changes here for the test cases may not be necessary.
I think that things like this may ultimately be resolved if it is not a human writing the details. Some ideas would be to allow some compilation of some part of the sympy code in more efficient form for the interpreter, but it would not turn out to be easy to be implement.
The other way would be using LRU cache for singleton registry, and make it skip the sympify for commonly used integers, But I'm not very sure the changes would be in totally forward direction.

oscarbenjamin commented:
I think that clarity and correctness should take a higher priority than speed when it comes to micro-optimisations like this. We should first identify/agree what is best and what is clearest. Then if the best way to do something isn't the fastest we should first try to make the best way faster. Only at the point where we have already optimised the best way but using something else offers a *measurable* improvement should we consider using code that is otherwise suboptimal purely for speed reasons.

Maybe if we want to use `Rational(1, 2)` everywhere instead of `S(1)/2` it would make sense to define `R = Rational` in the sympy namespace so it can be `R(1, 2)`. If we think that that is clearer and more "correct" then I'd be in favour of using that generally. On the other hand if we think that `S(1)/2` is really the "best" way to do this then maybe we can make it faster by speeding up S(int) or `Integer/Integer`?

smichr commented:
> R = Rational

I think it is frowned on to provide too many single-letter classes.

The only place that I am aware of large time savings is in the ntheory functions where calculations done with Python ints are much faster than those with Integers. But once you start making those numbers part of a symbolic expression...the savings is immaterial.:
```python
>>> t=time()
... for s in symbols('y:1000'):
...     ok=s + S(1)/2
... print(time()-t)
0.505486965179

>>> t=time()
... for s in symbols('z:1000'):
...     ok=s + Rational(1, 2)
... print(time()-t)
0.503242015839
```

sylee957 commented:
I'm personally not against merging some changes like this, if the speed improvement is legitimate.

I have only skimmed through 50 files, however, I think that there are some inconsistency of using `is_zero` vs `is S.Zero` in the changes.

Personally, I think that `.is_zero` can be used in place of `is S.Zero` since the original implementation would strongly assert that the object is a sympy object.
And I personally prefer using `.is_zero` for every cases possible, because it provides more space for the future allowing the algorithm to work with symbols, and also the slowdown can be negligible at as the assumptions are cached (But may only take an effect for the second call).

If there is no proper assumption predicate, then it can be fine with using `== Sqrt(2)`, but even for this case, I think that using `is` comparison may create some problems because there can still be a lot of old code laying around, using python numbers or having both sympy and python objects mixed in for the same attribute.

oscargus commented:
> I think that clarity and correctness

Can you point to any place where this is a problem with the new code?

Regarding Rational etc. 

```
In [2]: %timeit q = Quaternion(-S(7)/225, -S(1)/225, -S(1)/150, -S(2)/225)
15.5 µs ± 338 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [3]: %timeit q = Quaternion(Rational(-7, 225), Rational(-1, 225), Rational(-1, 150), Rational(-2, 225))
4.06 µs ± 38.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Apart from slightly longer code, is the second really worse in any other aspect?

(And I guess a reason that S.Half is introduced is really speed. Btw, it is better to multiply by S.Half than dividing by 2, but I have not really enforced this here.)

oscargus commented:
> There are some other concerns like exceeding 78-line limits

This is not really strictly enforce anyway in the code and I there are clearly instances here where the lines where earlier >78 characters and are now fixed.

> Personally, I think that .is_zero can be used in place of is S.Zero since the original implementation would strongly assert that the object is a sympy object.

The thing here is that `.is_zero` introduces the whole assumptions machinery, so there is a risk that an `is_zero` check require much more time compared to `is S.Zero`. However, compared to `== 0`, `== S.Zero`, and `== S(0)`, `is S.Zero` is faster and more pythonic, so those things shouldn't really be an issue.

In #17510 I have introduced double checks with `is S.Zero` early in the `eval` and `.is_zero` later on. As always with nested if-clauses it is a bit of speculation in which order to put them, but it probably makes sense to test with `is ...` early (as these are very simple checks) and possibly more complicated ones later.

But, yes, the code base is inconsistent in this respect as a whole and while there are some odd changes here, I have not really done that change systematically in this PR.

oscarbenjamin commented:
> > I think that clarity and correctness
> 
> Can you point to any place where this is a problem with the new code?

I just mean that the first priority should not be speed, especially when considering alternatives that are not semantically equivalent. In most cases `is_zero` is probably more correct than `== 0` in which case that should be the reason to use it.

oscarbenjamin commented:
> It seems like `is_zero` is clearly faster than `== S.Zero`, `== 0`, `is S.Zero`, etc. At least if it is a symbol or sympified constant. For expressions it varies

You have to be careful using timeit for with things that use caching. For a given object, only the first call to is_zero actually computes anything.

oscarbenjamin commented:
I've just rebased this. It was time consuming to rebase this so as @oscargus said above we need a decision on this quickly if we want it in.

I would have like to have had more discussion about the best style for the code but this is here now and makes a lot of small improvements. I don't think there is anything particularly controversial here.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -162,7 +162,7 @@ def function_range(f, symbol, domain):
 
     period = periodicity(f, symbol)
-    if period is S.Zero:
+    if period == S.Zero:
```
It looks like `S.Zero` is the only result from `periodicity` that could satisfy this.

jksuom commented:
I only checked a part of this but it seems ok to me. +1

oscarbenjamin commented:
Thanks @jksuom. I'm going to merge this before any more conflicts appear

oscarbenjamin commented:
Thanks @oscargus
